README
PostScript::Report version {{$version}}, released {{$date}}


PostScript::Report helps you generate nicely formatted reports using
PostScript. You do not need any knowledge of PostScript to use this
package (unless you want to create new Component types).



INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install



DEPENDENCIES

{{ $t->dependency_list }}



CHANGES
    Here's what's new in version {{$version}} of PostScript::Report:
    (See the file "Changes" for the full revision history.)

{{$changes}}



COPYRIGHT AND LICENSE

Copyright 2010 Christopher J. Madsen

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for PostScript-Report

0.13   NOT
	- Builder now uses Module::Runtime to load classes
	- Fix type library to be compatible with upcoming Moose 2.1200
	  (fixes RT#90697 & RT#90698)
	- Localize $@ before eval

0.12   2013-05-04
	- No functional changes; just updated metadata
	- Add Moose & MooseX::Types to configure_requires.
	  They're not actually required by Makefile.PL, but if they're
	  not installed, it skips the version check introduced in
	  version 0.11.  Since many CPAN testers don't install modules
	  normally, but just build an @INC containing only the listed
	  requirements, they always skipped the version check and
	  continued to send reports for that Moose bug.

0.11   2013-04-27
	- Clean up type library
	- Warn about conflict between MooseX::Types 0.32 and greater
	  with Moose versions before 2.0800 (fixes RT#77299)
	- Stabilize order of font creation.  Didn't affect printed
	  output, but generated code could be different between runs,
	  causing test failures on Perl 5.17.

0.10   2012-04-04
	- Use the function library provided by PostScript::File 2.20

0.09   2011-05-05
	- Use the improved API in PostScript::File 2.10
	- Fixed LinkField to avoid deprecation warnings in Perl 5.12

0.08   2010-07-20
	- Added first_footer attribute
	- Added "split" as a possible value for footer_align
	- Include more report attributes in dump output
	- Fixed bug in calculating required page count
	  (could cause report footer to overlap report)

0.07   2010-06-19
	- Added extra_styles & get_style, which allows Components to
	  inherit attributes without Containers and the Report needing
	  to know every possible attribute
	- Removed label_font from Report & Component
	  (It's only used by FieldTL, and can be handled by extra_styles.)

0.06   2010-03-26
	- Added the LinkField component
	- Added the metrics attribute to Font
	- Fixed Image to round calculated dimensions to nearest point
	  (A Component's height & width must be integers.)

0.05   2010-02-26
	- Use PostScript::File::Metrics (requires PostScript::File 2.00)
	  You no longer need Font::AFM to get word wrapping.
	- Call pstr as a method to take advantage of auto_hyphen
	  processing in PostScript::File 2.00.	You should read
	  Hyphens and Minus Signs in PostScript::File's documentation.
	- Fixed memory leak (Font should have used a weak reference)
	- Fixed Image to correctly initialize graphics state for EPS file
	- Added get__PostScript_File method for PostScript::Convert
	- Added examples/pdf_image.pl showing PostScript::Convert usage
	- Require Moose 0.90 instead of MooseX::AttributeHelpers

0.04   2009-10-29
	- You can now indicate which sides should have a border
	- You can now omit \%data and/or \@rows when calling run
	- Use cp1252 encoding (requires PostScript::File 1.05)
	- Fixed crash caused by a VBox with only 1 child

0.03   2009-10-28
	- Added background attribute to Component
	- Added detail_background attribute to Report
	- Added stripe & stripe_page to Builder
	- Add PageOrder: Ascend comment
	- Fixed bug in multiline FieldTL (was using wrong line height)
	- Warn in documentation against making padding_bottom too small

0.02   2009-10-22
	- Builder allows scalar ref as shortcut for Value::Constant
	  and _default overrides default_field_type in a container
	- Warn if a field's value does not exist.
	  Do not warn about existing but undefined values.
	- Added the ps_parameters attribute to Report
	- Updated documentation

0.01   2009-10-20
	- Initial release


examples/multiline_header.pl
#! /usr/bin/perl
#---------------------------------------------------------------------
# This example report is hereby placed in the public domain.
# You may copy from it freely.
#
# This is a fairly complex example for PostScript::Report.
# It includes a header field that spans two rows and uses word wrap.
#---------------------------------------------------------------------

use strict;
use warnings;

use PostScript::Report ();

# Define a Value::Constant with a blank string:
my $blank = \'';

my $desc = {
  fonts => {
    label     => 'Helvetica-6',
    text      => 'Helvetica-9',
    boldText  => 'Helvetica-Bold-9',
    pageNum   => 'Helvetica-8',
    bottomRow => 'Helvetica-6',
    disclaimerText => 'Helvetica-Bold-8',
  },

  font       => 'text',
  label_font => 'label',
  align      => 'center',

  padding_bottom => 4,
  padding_side   => 3,

  landscape     => 1,
  top_margin    => 25,
  left_margin   => 20,
  right_margin  => 20,
  bottom_margin => 25,
  row_height    => 22,

  default_field_type => 'FieldTL',

  # The report_header is one line with text fields left, center, & right.
  # All values are constant.
  report_header => [
    HBox => { border => 0,
              font => 'boldText',
              height => 12,
              padding_side   => 0,
              _default => 'Field', # override default_field_type
            },
    { width  => 200,
      align  => 'left',
      value  => \'AIRPLANE COMPONENT FIXING INC.' },
    { width  => 351,
      value  => \'WORK ORDER' },
    { width  => 200,
      align  => 'right',
      value  => \'F.A.A REPAIR STATION NO. L3PF428Q' },
  ],

  # The page_header is fairly complex.
  # The Material Type field spans rows 2 and 3.
  page_header => [
    # This HBox is row 1 of the header:
    [ { label => 'Customer Name:',
        value => 'custName',
        width => 160 },
      { label => 'Part Number Received:',
        value => 'partNumReceived',
        width => 146 },
      { label => 'Serial Number Received:',
        value => 'serialNumReceived',
        width => 156 },
      { label => 'Installed On:',
        value => 'installedOn',
        align => 'left',
        width => 130 },
      { label => 'Location:',
        value => 'location',
        width => 54 },
      { label => 'Work Order#:',
        value => 'workOrder',
        font  => 'boldText',
        width => 105 },
    ],
    # This HBox contains rows 2 & 3, because Material Type is 2 rows high:
    [
      # This VBox is the left hand side of rows 2 & 3:
      [
        # This HBox is the left hand side of row 2:
        [ { label => 'Part Description:',
            value => 'partDesc',
            width => 160 },
          { label => 'Part Number Returned:',
            value => 'partNumReturned',
            width => 146 },
          { label => 'Serial Number Returned:',
            value => 'serialNumReturned',
            width => 156 },
        ], # end row 2 left
        # This HBox is the left hand side of row 3:
        [ { label => 'Date Received:',
            value => 'dateReceived',
            width => 69 },
          { label => 'RO Due Date:',
            value => 'roDueDate',
            width => 91 },
          { label => 'Repair/Overhaul Per:',
            value => 'repairPer',
            align => 'left',
            width => 302 },
        ] ], # end rows 2 & 3 left
      # This field is two rows high:
      { label => 'Material Type:',
        value => 'materialType',
        align => 'left',
        height    => 44,
        multiline => 1,
        width => 130 },
      # This VBox is the right hand side of rows 2 & 3:
      [
        # Don't need an HBox for row 2 right; it's only one field:
        { label => 'Customer Order Number:',
          value => 'custOrderNum',
          width => 159 },
        # This HBox is the right hand side of row 3:
        [ { label => 'Part Verified By:',
            value => 'verifiedBy',
            width => 80 },
          { label => 'Revised Due Date:',
            value => 'revisedDueDate',
            width => 79 },
        ] ], # end rows 2 & 3 right
    ], # end rows 2 & 3
  ], # end page_header

  columns => {
    header => {
      font           => 'boldText',
      height         => 19,
      padding_bottom => 6,
    },
    detail => {
      height         => 19,
      padding_bottom => 6,
    },
    data => [
      [ 'SEQ#' => 29 ],
      [ 'STA#' => 40 ],
      [ 'REPAIR SCOPE' => 450, { align => 'left'}, { align => 'left'} ],
      [ MECHANIC => 73 ],
      [ INSPECTOR => 80 ],
      # The DATE column is filled in by hand after printing.  We'll
      # use a Spacer in the detail section so we don't need a blank
      # column in each row.  (Note: If this weren't the last column,
      # we'd also need to pass a fake value to the Spacer to tell the
      # builder not to increment the column number.)
      [ DATE => 79, undef, { _class => 'Spacer' } ],
    ],
  }, # end columns

  # The footer is also a bit complex:
  page_footer => [
    # We don't want a border around the whole footer:
    VBox => { border => 0 },
    # The first row is just one component:
    { _class => 'Field',
      font   => 'disclaimerText',
      value  => \'The component identified above was repaired/overhauled/inspected IAW current federal aviation regulations and in respect to that work, was found airworthy for return to service.',
    },
    # The second row is filled in by hand, so the values are blank:
    [ HBox => { border => 1 },
      { label => 'Inspector',
        value => $blank,
        width => 339 },
      { label => 'Final Inspection Stamp',
        value => $blank,
        width => 154 },
      { label => 'Date',
        value => $blank,
        width => 258 },
    ],
    # The third row includes the page number and checkboxes:
    [ HBox => {
        border => 1,
        height => 14,
        font => 'bottomRow',
        padding_side => 0,
        _default => 'Field',
      },
      # This HBox exists just to set parameters on its children:
      [ HBox => { border => 0, font  => 'pageNum' },
        { value => \'42410-1',
          width => 57 },
        { _class => 'Spacer',
          width  => 14 },
        { align => 'left',
          value => { _class => 'Page', value => 'Page(s): %n OF %t' },
          width => 377 },
      ],
      # This HBox exists just to set parameters on its children:
      [ HBox => { border => 0, padding_bottom => 6, padding_side => 2 },
        { align => 'right',
          value => \'ML',
          width => 10 },
        { align => 'right',
          value => \'FXN-',
          width => 53 },
        { align => 'left',
          value => 'FXN',
          width => 28 },
        { align => 'right',
          value => \'AP',
          width => 17 },
        { _class => 'Checkbox',
          value => 'AP',
          padding_bottom => 3,
          width => 20 },
        { align => 'right',
          value => \'BP',
          width => 20 },
        { _class => 'Checkbox',
          value => 'BP',
          padding_bottom => 3,
          width => 20 },
        { align => 'right',
          value => \'Photo',
          width => 45 },
        { align => 'left',
          value => 'photo',
          width => 28 },
        { align => 'right',
          value => \'QTY',
          width => 35 },
        { align => 'left',
          value => 'qty',
          width => 27 },
      ],
    ], # end third row of page_footer
  ], # end page_footer
};

my $data = {
  'custName'          => 'IMAGINARY AIRWAYS',
  'partNumReceived'   => '957X1427-3',
  'serialNumReceived' => 'N/A',
  'installedOn'       => '797',
  'location'          => 'A1',
  'workOrder'         => '68452-8',
  'partDesc'          => 'TURBOFAN',
  'partNumReturned'   => '957X1427-3',
  'serialNumReturned' => 'N/A',
  'materialType'      => 'FOO BAR 123',
  'custOrderNum'      => '8452647',
  'dateReceived'      => '05/06/2009',
  'roDueDate'         => '05/06/2009',
  'repairPer'         => 'REPAIR PER B797 CMM 47-42-96 REV 40 DATED 07MAY2009',
  'verifiedBy'        => '951',
  'revisedDueDate'    => '',
  AP                  => 0,
  BP                  => 1,
  FXN                 => 0,
  photo               => 6,
  qty                 => 1,
};

my $rows = [
  [ 1, 'I1', 'INSPECT, PN & SN VERIFIED', 'XXXXXXXX', '' ],
  [ 2, 'I1', 'ADDITIONAL DATA USED: BOEING ASSEMBLY DRAWING 589X1674, GDR9726', 'XXXXXXXX', '' ],
  [ 3, 'I1', 'Lorem ipsum dolor sit amet, consectetur adipisicing elit', 'XXXXXXXX', '' ],
  [ 4, 'I1', 'Lorem ipsum dolor sit amet', 'XXXXXXXX', '' ],
  [ 5, 'I1', 'Lorem ipsum dolor sit amet', 'XXXXXXXX', '' ],
  [ 6, 'I1', 'Lorem ipsum dolor sit amet', 'XXXXXXXX', '' ],
  [ 7, 'I1', 'Lorem ipsum dolor sit amet', 'XXXXXXXX', '' ],
  [ 8, 'S1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 9, 'W1', 'Lorem ipsum dolor sit amet', '', '' ],
  [ 10, 'W1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 11, 'W1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 12, 'W1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 13, 'W1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 14, 'W1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 15, 'W1', 'Lorem ipsum dolor sit amet', '', '' ],
  [ 16, 'P1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 17, 'P1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 18, 'P1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 19, 'P1', 'Lorem ipsum dolor sit amet', '', '' ],
  [ 20, 'P1', 'Lorem ipsum dolor sit amet', '', '' ],
  [ 21, 'F1', 'Lorem ipsum dolor sit amet', 'XXXXXXXX', '' ],
  [ 22, 'P1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 23, 'P1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 24, 'S1', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 25, 'SR', 'Lorem ipsum dolor sit amet', '', 'XXXXXXXX' ],
  [ 26, 'I1', 'Lorem ipsum dolor sit amet', 'XXXXXXXX', '' ],
];

my $rpt = PostScript::Report->build($desc);

$rpt->run($data, $rows)->output("multiline_header.ps");

# $rpt->dump;
